stdlib/pervasives.ml

```
@@ -369,7 +373,7 @@ external seek_out : out_channel -> int -> unit = "caml_ml_seek_out"
 external pos_out : out_channel -> int = "caml_ml_pos_out"
 external out_channel_length : out_channel -> int = "caml_ml_channel_size"
 external close_out_channel : out_channel -> unit = "caml_ml_close_channel"
-let close_out oc = flush oc; close_out_channel oc
+let close_out oc = (try flush oc with _ -> ()); close_out_channel oc
```

Buffered data could be silently lost if `flush` fails.  That makes this particular change seem dangerous.

---

@hcarty Note that this patch only makes `stdlib` and `threads` consistent.  In this case I chose the semantics as already implemented in `threads`, which I considered saner.  It will close the channel for sure, otherwise we would be leaking file descriptors.  If a user absolutely wants to make sure that their data has been flushed, they need to explicitly call `flush` themselves before calling `close_out`, thus potentially raising `Sys_error` exceptions.  `flush_all` also suppresses such exceptions, because the user could not possibly know anyway which channel was broken.

---

@mmottl Scope of a surrounding `try`/`with` would affect whether there is a leak or not here.

Regardless, this change should be reflected in the changelog and documentation to call out the requirement for an explicit call to `flush` to guarantee flushing of buffered output.

---

@hcarty Ok, I have looked at the documentation and it actually already specifies that `close_out` may raise even while flushing and before closing.  Users who don't like that behavior should use `close_out_noerr`, which completely ignores errors.  This means that the previous implementation in `threads` was not compliant with the manual while the `stdlib` was.

I have just submitted a patch to make the current `stdlib` behavior the default with both implementations.  They must definitely behave the same for that function.  It would be very confusing for users if that wasn't the case.  If some thread-related library needed the previous closing behavior, that code would have to be changed to use `close_out_noerr`.  I have looked, but haven't seen any such code in the distribution.  The semantics as specified in the manual should definitely prevail.